go/kbfs/search/indexer.go

```diff
@@ -0,0 +1,364 @@
+	go func() {
+		select {
+		case <-waitCh:
+		case <-i.shutdownCh:
+			return
+		}
+
+		m := tlfMessage{tlfID, rev, keybase1.FolderSyncMode_ENABLED}
+		select {
+		case i.tlfCh <- m:
+		default:
+			i.log.CDebugf(
+				context.Background(), "Couldn't send TLF message for %s/%d")
+		}
+	}()
+}
+
+// SyncModeChanged implements the libkbfs.SyncedTlfObserver interface
+// for Indexer.
+func (i *Indexer) SyncModeChanged(
+	ctx context.Context, tlfID tlf.ID, newMode keybase1.FolderSyncMode) {
+	i.log.CDebugf(ctx, "Sync mode changed for %s to %s", tlfID, newMode)
+	m := tlfMessage{tlfID, kbfsmd.RevisionUninitialized, newMode}
+	select {
+	case i.tlfCh <- m:
+	default:
+		i.log.CDebugf(
+			context.Background(), "Couldn't send TLF message for %s/%d")
+	}
+}
+
+var _ libkbfs.SyncedTlfObserver = (*Indexer)(nil)
+
+func (i *Indexer) fsForRev(
+	ctx context.Context, tlfID tlf.ID, rev kbfsmd.Revision) (*libfs.FS, error) {
+	if rev == kbfsmd.RevisionUninitialized {
+		return nil, errors.New("No revision provided")
+	}
+	branch := data.MakeRevBranchName(rev)
+
+	md, err := i.config.MDOps().GetForTLF(ctx, tlfID, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	h := md.GetTlfHandle()
+	return libfs.NewReadonlyFS(
+		ctx, i.config, h, branch, "", "", keybase1.MDPriorityNormal)
+}
+
+func (i *Indexer) loop(ctx context.Context) {
+	i.log.CDebugf(ctx, "Starting indexing loop")
+	defer i.log.CDebugf(ctx, "Ending index loop")
+
+	i.remoteStatus.Init(ctx, i.log, i.config, i)
+
+outerLoop:
+	for {
+		err := i.loadIndex(ctx)
+		if err != nil {
+			i.log.CDebugf(ctx, "Couldn't load index: %+v", err)
+		}
+
+		state := keybase1.MobileAppState_FOREGROUND
+		kbCtx := i.config.KbContext()
+		for {
+			select {
+			case <-i.userChangedCh:
+				// Re-load the index on each login/logout event.
+				i.log.CDebugf(ctx, "User changed")
+				continue outerLoop
+			case state = <-kbCtx.NextAppStateUpdate(&state):
+				// TODO(HOTPOT-1494): once we are doing actual
+				// indexing in a separate goroutine, pause/unpause it
+				// via a channel send from here.
+				for state != keybase1.MobileAppState_FOREGROUND {
+					i.log.CDebugf(ctx,
+						"Pausing indexing while not foregrounded: state=%s",
+						state)
+					state = <-kbCtx.NextAppStateUpdate(&state)
+				}
+				i.log.CDebugf(ctx, "Resuming indexing while foregrounded")
+				continue
+			case m := <-i.tlfCh:
+				ctx := i.makeContext(ctx)
+				i.log.CDebugf(ctx, "Received TLF message for %s", m.tlfID)
+
+				// Figure out which revision to lock to, for this
+				// indexing scan.
+				var rev kbfsmd.Revision
+				if m.rev == kbfsmd.RevisionUninitialized {
+					// TODO(HOTPOT-1504) -- remove indexing if the
+					// mode is no longer synced.
+					continue
+				}
+
+				_, err := i.fsForRev(ctx, m.tlfID, rev)
+				if err != nil {
+					i.log.CDebugf(ctx, "Error making FS: %+v", err)
+					continue
+				}
+
+				// TODO(HOTPOT-1494, HOTPOT-1495): initiate processing
+				// pass, using the FS made above.
```

Comment:
Does this mean any indexing has to finish within two weeks? Otherwise some of the block referenced by this revision but deleted later might get GC'ed before this revision's indexing finishes.

Reply:
My plan was that if blocks become unfetchable, the indexing will fail, and it will likely have to start over at the latest revision and do a full re-index.